gobblin-iceberg/src/main/java/org/apache/gobblin/iceberg/writer/IcebergMetadataWriter.java

```diff
@@ -947,9 +938,15 @@ private long computeCompletenessWatermark(String table, SortedSet<ZonedDateTime>
         if (timestampDT.isAfter(prevWatermarkDT)
             && TimeIterator.durationBetween(prevWatermarkDT, now, granularity) > 0) {
           long timestampMillis = timestampDT.toInstant().toEpochMilli();
-          if (auditCountVerifier.get().isComplete(table,
-              TimeIterator.dec(timestampDT, granularity, 1).toInstant().toEpochMilli(), timestampMillis)) {
+          ZonedDateTime auditCountCheckLowerBoundDT = TimeIterator.dec(timestampDT, granularity, 1);
+          if (auditCountVerifier.get().isComplete(topicName,
+                  auditCountCheckLowerBoundDT.toInstant().toEpochMilli(), timestampMillis)) {
             completionWatermark = timestampMillis;
+            // Also persist this into State object to share this with other MetadataWriters
+            // we enforce ourselves to always use lower-cased table name here
+            String tableNameLowerCased = tableName.toLowerCase(Locale.ROOT);
+            this.state.tableToCompletedPartitions.putIfAbsent(tableNameLowerCased, Sets.newHashSet());
+            this.state.tableToCompletedPartitions.get(tableNameLowerCased).add(timestampDT.format(HOURLY_DATEPARTITION_FORMAT));
```

Comment:
If we have jumped in the watermark, i.e. we directly update the watermark from 1 am to 8 am, will you only have the  8 am partition in the completed partition list and miss partitions from 2 am to 7 am?

Reply:
Yeah, if we just use whatever in the `state` as is, then we will miss publishing the completion event for 2-7 am. But I think we can have a way to maintain the previously completed partition in the state for jasper to compute the delta, so upon the 8 am watermark, jasper can publish the completion event for 2-8 am altogether. Or, do you think we have a better alternative approach to this?

But this will mean a huge SLA delay for downstream, does gobblin guarantee a SLA on the watermark?

Reply:
We do guarantee SLA in most cases, but there are exceptions and GCNs can happen downstream which will cause these delays. My suggestion is to have Jasper expose the last completed partition and Jasper writer to compute the delta here as well. But it's just a reminder here and no need to address in this PR, maybe have one comment to indicate this scenario

Reply:
ok, sounds good, i will talk to the team to decide how to manage the gap and jump on watermark case like this.